language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -315,3 +362,34 @@ onHead :: (a -> a) -> [a] -> [a]
 onHead f = \case
     [] -> []
     x:xs -> f x:xs
+
+onLast :: (a -> a) -> [a] -> [a]
+onLast f = \case
+  [] -> []
+  [l] -> [f l]
+  x : xs -> x : onLast f xs
+
+-- A type such as
+--   data Q = C { x: Int, y: Text }| G { z: Bool }
+-- has a DAML-LF representation like,
+--   record Q.C = { x: Int, y: String }
+--   record Q.G = { z: Bool }
+--   variant Q = C Q.C | G Q.G
+-- Use this function to get the definitions of 'Q.C' and 'Q.G' given
+-- 'Q'.
+assocDefDataTypes :: Module -> T.Text -> [((T.Text, T.Text), DefDataType)]
+assocDefDataTypes m c =
+  [((sup, sub ), def) | def <- defDataTypes m
+    , [sup, sub] <- [unTypeConName (dataTypeCon def)], sup == c]
+
+-- Extract the serialization code associated with a data type
+-- definition.
+serFromDef :: ([T.Text] -> c)
+           -> (T.Text, T.Text)
+           -> PackageId -> Module -> NM.NameMap Template
+           -> DefDataType -> c
+serFromDef f (c1, c2) curPkgId mod tpls = f . snd . fst . genDefDataType curPkgId (c1 `dot` c2) mod tpls
```

Comment:
This function is so simple, we should inline it, particularly since it's quite hard to come up with a good descriptive name for it.

Reply:
There are two call sites, that's why I factored it out - it's a beast. What I could do is make it local to `genDefDataType`. Would that do?

Reply:
Yes, that would do.